### PR TITLE
Limit submission rate for a question

### LINKED ISSUE #1195: Rate limit submissions
Add an option to assessment questions to rate limit submissions. It's probably better to have it set per-question and per-assessment, not just in the question itself.

### PR DESCRIPTION
**Pull request by jonatanschroeder**
Resolves #1195 

- [x] Schema updates
- [x] Database updates for configuration
- [x] Question instance UI updates
- [X] Lib updates that block open instance questions from grading a submission based on rate
- [x] Assessment overview UI updates
- [X] Handling of assessment close (grade even if faster than rate?)
- *** Add Grade all/Close all button on assessment instance list
   - will be added in a separate PR
- [x] Create tests
- [x] Documentation

### CODE REVIEW

**Comment by jonatanschroeder** [comment]
@mwest1066  Lib is updated and does the necessary checks in the background, so now it's a matter of updating UI, documentation and tests. Regrading is not affected, since it uses a different lib altogether. For UI:

* The instance question UI has a preliminary update, but there will be tons of conflicts with #3601 so I'll wait until you merge that before moving on with that one.
* I like the idea of having a countdown-like interface, but I believe it would be a stretch to use the same partial as the time limit one, since that one has several processes that are specific to that use-case (use of "danger" colors closer to the expiry of the time, submitting a form once the time expires, etc.). I might separate the UI aspect from the functionality aspect of that partial, then use that instead (I need to experiment before committing to something).
* Should there be some indication in the UI for the user, before the submission, that there is a submission rate in effect? Maybe similar to the "additional attempts" note used in #3601? We can always send the responsibility of explaining that to the instructor, in question.html, but from experience this doesn't work well.
* For the assessment overview, right now it properly skips questions blocked due to submission rate from being graded (since it uses the same lib function), but the interface doesn't provide any indication that there is a question that can't be graded due to rate. Any ideas on how to present that?
  * The easiest solution would be to add a note after the "Grade XX answers" button that says something to the effect "at least one saved submission can't be graded because its submission rate has not been reached yet."
  * Another option is to change the counter in this button to only include gradable questions, but that would also need to include a statement on why a question with status "saved" cannot be graded.
  * We can also include an icon next to "saved" status in the question list that opens a popover explaining the submission rate.
* Do we need to provide instructors with some view of the blocked rate in effect for a student? Maybe in the assessment instance log?

**Comment by mwest1066** [comment]
> * I like the idea of having a countdown-like interface, but I believe it would be a stretch to use the same partial as the time limit one, since that one has several processes that are specific to that use-case (use of "danger" colors closer to the expiry of the time, submitting a form once the time expires, etc.). I might separate the UI aspect from the functionality aspect of that partial, then use that instead (I need to experiment before committing to something).

Makes sense. If it's possible to abstract the actual countdown code then that would be great, because this code is slightly subtle and I'd rather not have multiple versions of it. But do whatever you think is best.

> * Should there be some indication in the UI for the user, before the submission, that there is a submission rate in effect? Maybe similar to the "additional attempts" note used in #3601? We can always send the responsibility of explaining that to the instructor, in question.html, but from experience this doesn't work well.

I agree that we should never leave this sort of thing up to the question author. They will always forget or write something unclear 😄. Perhaps we could add a note to the question footer like "Can only be graded once every X minutes"?

> * For the assessment overview, right now it properly skips questions blocked due to submission rate from being graded (since it uses the same lib function), but the interface doesn't provide any indication that there is a question that can't be graded due to rate. Any ideas on how to present that?
>   * The easiest solution would be to add a note after the "Grade XX answers" button that says something to the effect "at least one saved submission can't be graded because its submission rate has not been reached yet."
>   * Another option is to change the counter in this button to only include gradable questions, but that would also need to include a statement on why a question with status "saved" cannot be graded.
>   * We can also include an icon next to "saved" status in the question list that opens a popover explaining the submission rate.

I like the idea of putting an icon (maybe a clock?) next to the "saved" status for questions with a limit in place. And then we can also include the statement about "at least one ... reached yet (marked with [clock-icon] above). Reload the page to update.". By explicitly referencing [clock-icon] I hope it will be clear what is happening. To be clear, I don't think we should live-update these icons. Or should we? Seems like lots of work for minimal gain.

> * Do we need to provide instructors with some view of the blocked rate in effect for a student? Maybe in the assessment instance log?

You mean a real-time view? Or a log for after the assessment is over? I'm not understanding what we would show in either case?

**Comment by jonatanschroeder** [comment]
> > * Do we need to provide instructors with some view of the blocked rate in effect for a student? Maybe in the assessment instance log?
> 
> You mean a real-time view? Or a log for after the assessment is over? I'm not understanding what we would show in either case?

I was thinking more in a sense of helping students figure out why they can't submit for grading, but I guess that's not a big deal now. Instructors can check the student view if needed, and if something else comes up after implementation we can have it as a separate PR.

> I like the idea of putting an icon (maybe a clock?) next to the "saved" status for questions with a limit in place.

I have a stopwatch right now, which looks cute, but is possibly too discrete. Should I maybe add some color or increase the size?

Another question: if a student saves the answer when credit is 100, but then grades when credit is 0, which credit will apply? Which credit should apply? What if they save but never grade (thinking of Homework-style, where you could potentially never close the assessment)? Could there be a mechanism to maybe grade all saved assessments at the deadline?

**Comment by jonatanschroeder** [comment]
image

Any suggestions? Bigger, different color, etc.?

**Comment by mwest1066** [comment]
> Any suggestions? Bigger, different color, etc.?

Somehow that icon looks more like a hand grenade than a clock 😄 

The vertical alignment looks off? Perhaps an `align-middle` would help? https://getbootstrap.com/docs/4.5/utilities/vertical-align/

**Comment by mwest1066** [comment]
Ah, I think you are using `fa-stopwatch`. I have a slight preference for `<i class="far fa-clock"></i>` because it looks more like a clock to me. I don't think I would have guessed what the stopwatch icon is!

**Comment by jonatanschroeder** [comment]
I don't like clocks for countdowns, I replaced it with an hourglass (`fa fa-hourglass-half`)...

For the Homework overview page, it doesn't have a "Grade XX saved answers", nor does it show "saved" status. I imagine keeping that information in the question only, and not messing with the overview page, may be the way to go.

Repeating it here in case you missed it: 
> if a student saves the answer when credit is 100, but then grades when credit is 0, which credit will apply? Which credit should apply? What if they save but never grade (thinking of Homework-style, where you could potentially never close the assessment)? Could there be a mechanism to maybe grade all saved assessments at the deadline?

Would it be viable/acceptable to have some way to, from the list of submissions under a question/instance question, have some button that grades that specific submission (could be limited to the last one) instead of requiring a "Save & Grade" to create a new submission? If such a button exists, we could use the credit at the time of saving instead of grading, if you think that's appropriate.

**Comment by jonatanschroeder** [comment]
I'm working on the countdown partial assuming #3386 will be approved. I'll keep an eye for changes there so they're properly applied in this code.

**Comment by mwest1066** [comment]
> I don't like clocks for countdowns, I replaced it with an hourglass (`fa fa-hourglass-half`)...

Looks great!

> For the Homework overview page, it doesn't have a "Grade XX saved answers", nor does it show "saved" status. I imagine keeping that information in the question only, and not messing with the overview page, may be the way to go.

Perfect.

> Repeating it here in case you missed it:
> 
> > if a student saves the answer when credit is 100, but then grades when credit is 0, which credit will apply? Which credit should apply? What if they save but never grade (thinking of Homework-style, where you could potentially never close the assessment)? Could there be a mechanism to maybe grade all saved assessments at the deadline?

I believe (but I haven't checked) that when we create the submission we save the current credit as `submissions.credit` and this is what is used for any grading of that submission. This applies even if the grading occurs much later (e.g., a slow external grader, automated grading on assessment close, etc).

If the student saves but never grades on a homework then they never get the points. In practice we've never found this to be a problem. Are you thinking that this could be a problem with rate-limited grading? In general I'm not very enthusiastic about delayed grading. It makes everything so much more complicated and I feel that it's generally hostile to students. What's the use case you are thinking about here?

> Would it be viable/acceptable to have some way to, from the list of submissions under a question/instance question, have some button that grades that specific submission (could be limited to the last one) instead of requiring a "Save & Grade" to create a new submission? If such a button exists, we could use the credit at the time of saving instead of grading, if you think that's appropriate.

I guess we could do that? But I'm not understanding what's the advantage of this?

**Comment by mwest1066** [comment]
> I'm working on the countdown partial assuming #3386 will be approved. I'll keep an eye for changes there so they're properly applied in this code.

I just shifted this to "Review in progress" to get it on the queue.

**Comment by jonatanschroeder** [comment]
> > Would it be viable/acceptable to have some way to, from the list of submissions under a question/instance question, have some button that grades that specific submission (could be limited to the last one) instead of requiring a "Save & Grade" to create a new submission? If such a button exists, we could use the credit at the time of saving instead of grading, if you think that's appropriate.
> 
> I guess we could do that? But I'm not understanding what's the advantage of this?

Use case (which I've seen in the past in courses where we had a rate limit): Homework-type assessment, deadline is 11:59:59, with a limit of one submission per 30 minutes. At 11:45, student submits what they expect to be their final version, but the autograder finds a bug that is easy to fix. They can save at 11:55, but will only be able to grade at 12:15, which is after the deadline. So if they can grade the 11:55 submission, they get the credit, but if they save & grade after that, they don't get the credit.

An alternative is to add a Close button in the Homework assessment overview that would grade all assessments and block future submissions. Another option is to add (in a future PR) the "Close All" button in the #2750 interface, and instructors select it after assessments are done. This would trigger a grade in all assessments. It is necessary because it's a Homework assessment, which doesn't typically include a Close button.

**Comment by mwest1066** [comment]
> Use case (which I've seen in the past in courses where we had a rate limit): Homework-type assessment, deadline is 11:59:59, with a limit of one submission per 30 minutes. At 11:45, student submits what they expect to be their final version, but the autograder finds a bug that is easy to fix. They can save at 11:55, but will only be able to grade at 12:15, which is after the deadline. So if they can grade the 11:55 submission, they get the credit, but if they save & grade after that, they don't get the credit.

Ok, thanks for reminding me about this! I still think rate-limiting is a bad idea, but I know lots of people want to do it so we should support it in good way 😄

> An alternative is to add a Close button in the Homework assessment overview that would grade all assessments and block future submissions. Another option is to add (in a future PR) the "Close All" button in the #2750 interface, and instructors select it after assessments are done. This would trigger a grade in all assessments. It is necessary because it's a Homework assessment, which doesn't typically include a Close button.

I like all these ideas! I'd be very happy to have instructor buttons for "Grade any outstanding submissions" and "Close and grade". These seem clearly good.

I'm also thinking about #3347 and how we can make Exams and Homeworks behave more the same.

I propose:

* We try and eliminate "closing" of assessments as much as possible. Maybe we can completely get rid of it. The only time we really need it at the moment is when a timer expires, and we could just check explicitly for an expired timer (and block access) rather than separately closing the assessment instance. We eliminate the `autoClose` option for Exams.
* For all assessments we automatically grade any outstanding submissions after the deadline/timer has expired. To do this, we add a flag `assessment_instances.has_ungraded_submissions BOOLEAN`. Every time we save a submission for this assessment instance, we set `has_ungraded_submissions := true`. We repurpose `cron/autoFinishExams` to search for any assessment instances that have `has_ungraded_submissions = true` and which are not accessible to the student, grade them, and set `has_ungraded_submissions := false`.
* On the instructorAssessmentInstances page we add a column to show `has_ungraded_submissions`. We also add a button to "Grade all ungraded submissions".
* Getting rid of "closing" would require implementing #3616 so that instructors can give students access to view (but not edit) their exams after they are over, even without the concept of "closing".

What do you think? Can we get rid of closing and move to a world where we always grade last-minute submissions?

**Comment by jonatanschroeder** [comment]
These sounds like great ideas, but I'd propose moving the elimination of "Close" to a separate PR/Issue. As for the implementation, I'd be in favour of having the autoFinishExams grade submissions/instance questions, not assessments. That would mean we don't need to keep the flag with cached information, as caching is prone to errors. The autoFinishExams would grade instance questions with status "saved" and whose assessment has expired (go up instead of down).

For now, I'll try to add "Close all" to #2750, and maybe add a "Grade"/"Grade all" option as well (these are exactly the same library function, only with a `false` instead of a `true` in one of the arguments).

**Comment by mwest1066** [comment]
> These sounds like great ideas, but I'd propose moving the elimination of "Close" to a separate PR/Issue. As for the implementation, I'd be in favour of having the autoFinishExams grade submissions/instance questions, not assessments. That would mean we don't need to keep the flag with cached information, as caching is prone to errors. The autoFinishExams would grade instance questions with status "saved" and whose assessment has expired (go up instead of down).

Oh yes, sorry, I didn't mean to imply that we should make any of these changes as part of this PR! 😄

I'll add a new Issue to document this.

> For now, I'll try to add "Close all" to #2750, and maybe add a "Grade"/"Grade all" option as well (these are exactly the same library function, only with a `false` instead of a `true` in one of the arguments).

Excellent!

**Comment by mwest1066** [comment]
> I'll add a new Issue to document this.

See #3622

**Comment by jonatanschroeder** [comment]
image

Thoughts on UI?

**Comment by mwest1066** [comment]
> Thoughts on UI?

Nice! Maybe "Grading possible in XX min", to sound more positive? 😄

Add a bit more vertical space below the Save&Grade button?

Add a question-mark icon next to "Can only be graded..." to explain that saving is always possible, and to explain what will happen to saved answers after the deadline passes (currently nothing, but we will autograde according to #3622).

**Comment by jonatanschroeder** [comment]
@mwest1066 before I move on to documentation and tests: I added the setting as `submissionRateLimitMin`, and can be set at the assessment top level, in a zone, in a question, or as an alternative for a question set. All 4 set the same field `submission_rate_limit_min` in `assessment_question`, and the first 3 are just defaults for the fourth. Is the name/structure ok or do you want something shorter/more limited?

**Comment by mwest1066** [comment]
> @mwest1066 before I move on to documentation and tests: I added the setting as `submissionRateLimitMin`, and can be set at the assessment top level, in a zone, in a question, or as an alternative for a question set. All 4 set the same field `submission_rate_limit_min` in `assessment_question`, and the first 3 are just defaults for the fourth. Is the name/structure ok or do you want something shorter/more limited?

Based on our experiences with `timeLimitMin`, I actually think we should use `Minutes` rather than `Min`. Many people thought that `Min` was short for `Minimum` and didn't understand what the units were. I don't like the fact that `submissionRateLimitMinutes` is so long, but I'd rather be long than unclear.

**Comment by jonatanschroeder** [comment]
> > @mwest1066 before I move on to documentation and tests: I added the setting as `submissionRateLimitMin`, and can be set at the assessment top level, in a zone, in a question, or as an alternative for a question set. All 4 set the same field `submission_rate_limit_min` in `assessment_question`, and the first 3 are just defaults for the fourth. Is the name/structure ok or do you want something shorter/more limited?
> 
> Based on our experiences with `timeLimitMin`, I actually think we should use `Minutes` rather than `Min`. Many people thought that `Min` was short for `Minimum` and didn't understand what the units were. I don't like the fact that `submissionRateLimitMinutes` is so long, but I'd rather be long than unclear.

Maybe submissionRateMinutes? Or gradeRateMinutes?

**Comment by mwest1066** [comment]
> Maybe submissionRateMinutes? Or gradeRateMinutes?

Oh yes, good point, it's actually grading that we are rate limiting. `gradeRateMinutes` sounds good. Or how about `gradeWaitMinutes` to emphasize that students need to wait this time before they can grade again?

**Comment by jonatanschroeder** [comment]
> > Maybe submissionRateMinutes? Or gradeRateMinutes?
> 
> Oh yes, good point, it's actually grading that we are rate limiting. `gradeRateMinutes` sounds good. Or how about `gradeWaitMinutes` to emphasize that students need to wait this time before they can grade again?

I don't like wait, it sounds like they have to submit then wait to grade. Rate sounds more like it's related to the previous occurrence.

**Comment by mwest1066** [comment]
> I don't like wait, it sounds like they have to submit then wait to grade. Rate sounds more like it's related to the previous occurrence.

That's fair, I'm fine with `gradeRateMinutes`.

File: sprocs/instance_questions_next_allowed_grade.sql
Comment (jonatanschroeder) [inline sprocs/instance_questions_next_allowed_grade.sql:6]: This will be removed, as it's only used in one place and requires additional queries to get the timezone. The computation of this date may be replaced by the interval as a text (e.g., 5 minutes, or 30 seconds).

File: pages/partials/examQuestionScore.ejs
Comment (jonatanschroeder) [inline pages/partials/examQuestionScore.ejs:21]: Replace the formatted date with the interval.

**Comment by jonatanschroeder** [comment]
@mwest1066 @nwalters512 it'd be great to have this working by the time we release the first assignment, mid-next week. Do you think this is possible?

File: pages/partials/examQuestionScore.ejs
Comment (mwest1066) [inline pages/partials/examQuestionScore.ejs:28]: How does this work if there are multiple questions with different timeouts? I think we should either do nothing (that is, remove this script) or do it properly (with unique IDs for each question, etc).

File: pages/partials/submitRateFooter.ejs
Comment (mwest1066) [inline pages/partials/submitRateFooter.ejs:20]: I'm trying to use "your answer" rather than "submission" when talking to students. I felt like the details about what instructors can do are not really helpful for students?
```suggestion
    <a class="btn btn-xs" data-toggle="popover" data-html="true" data-content="<p>Once you have clicked <strong>Save &amp; Grade</strong>, you will need to wait <%= assessment_question.grade_rate_minutes %> minute<% if (assessment_question.grade_rate_minutes > 1) { %>s<% } %> before you can grade again.</p><p>You can still save your answer as frequently as you like. <% if (question_context == "student_exam") { %>If your assessment ends before your last saved answer is graded, it will be automatically graded for you.<% } %></p>" data-placement="auto" data-trigger="focus" data-container="body" tabindex="0">
```

File: pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.ejs
Comment (mwest1066) [inline pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.ejs:243]:
```suggestion
              <li>There are <%= suspendedSavedAnswers %> saved answers that cannot be graded yet due to grading rate limits. They are marked with the <i class="fa fa-hourglass-half"></i> icon above. Reload this page to update this information.</li>
              <% } else if (suspendedSavedAnswers == 1) { %>
              <li>There is one saved answer that cannot be graded yet due to a grading rate limit. It is marked with the <i class="fa fa-hourglass-half"></i> icon above. Reload this page to update this information.</li>
```

File: pages/partials/examQuestionScore.ejs
Comment (mwest1066) [reply pages/partials/examQuestionScore.ejs:28]: Thinking about it more, I think we should just remove this. We aren't going to update the text that says "XX questions are rate limited", so things will get confusing in any case. I suggested a change to the information text to explicitly say that a page reload is needed for updates.

At some point we can rewrite this page with react and have everything update in real time 😄

File: pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.js
Comment (mwest1066) [inline pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.js:82]: Nit: I'll preempt @nwalters512 by requesting braces 😄

File: schemas/schemas/infoAssessment.json
Comment (mwest1066) [inline schemas/schemas/infoAssessment.json:106]:
```suggestion
            "description": "Minimum amount of time (in minutes) between graded submissions to the same question.",
```

File: schemas/schemas/infoAssessment.json
Comment (mwest1066) [inline schemas/schemas/infoAssessment.json:243]:
```suggestion
                    "description": "Minimum amount of time (in minutes) between graded submissions to the same question.",
```

File: schemas/schemas/infoAssessment.json
Comment (mwest1066) [inline schemas/schemas/infoAssessment.json:291]:
```suggestion
                    "description": "Minimum amount of time (in minutes) between graded submissions to the same question.",
```

File: schemas/schemas/infoAssessment.json
Comment (mwest1066) [inline schemas/schemas/infoAssessment.json:322]:
```suggestion
                    "description": "Minimum amount of time (in minutes) between graded submissions to the same question.",
```

File: sprocs/instance_questions_next_allowed_grade.sql
Comment (mwest1066) [inline sprocs/instance_questions_next_allowed_grade.sql:4]: We conventionally call these by their full names in argument lists.
```suggestion
        IN instance_question_id BIGINT,
```

File: sprocs/instance_questions_next_allowed_grade.sql
Comment (mwest1066) [inline sprocs/instance_questions_next_allowed_grade.sql:12]:
```suggestion
        MAX(gj.date + make_interval(mins => aq.grade_rate_minutes))
```

**Comment by mwest1066** [comment]
> @mwest1066 @nwalters512 it'd be great to have this working by the time we release the first assignment, mid-next week. Do you think this is possible?

Should be fine. I need to do the server-python update first. Don't hesitate to remind me about this.

When exactly do you need it deployed?

**Comment by mwest1066** [comment]
> Preferably by Tuesday next week (Jan 12), though earlier is better, gives us time to test it properly.

Ok, I'll try and get a deploy later this week. Don't hesitate to remind me.

File: pages/partials/examQuestionScore.ejs
Comment (jonatanschroeder) [reply pages/partials/examQuestionScore.ejs:28]: The issue is that this partial is also used in the instance question page, though it's easy enough to add this update in the function that is called when the countdown ends.

File: database/tables/assessment_questions.pg
Comment (jonatanschroeder) [reply database/tables/assessment_questions.pg:16]: No particular reason, should we allow half a minute? Autograders sometimes take longer than that to complete...

Comment (jonatanschroeder) [reply database/tables/assessment_questions.pg]: I changed it to allow double precision. This is the last change, everything else is resolved.

File: sprocs/instance_questions_next_allowed_grade.sql
Comment (jonatanschroeder) [reply sprocs/instance_questions_next_allowed_grade.sql]: I had to switch back, `make_interval` doesn't support double precision.

**Comment by mwest1066** [comment]
And I forgot to ask about a news item. Want to add one into this PR?

**Comment by jonatanschroeder** [comment]
> And I forgot to ask about a news item. Want to add one into this PR?

I can add one. Do you want screenshots or just the explanation?

**Comment by mwest1066** [comment]
> I can add one. Do you want screenshots or just the explanation?

Screenshots! Everyone loves an image 😄

**Comment by mwest1066** [comment]
@jonatanschroeder I edited the news item so that it's now in a more "action oriented" order:

1. What is this feature?
2. How do we configure it?
3. Why do we want it?

Please feel free to edit anything as always!

**Review by mwest1066 — APPROVED** [review]

**Comment by mwest1066** [comment]
> Please feel free to edit anything as always!

And tell me when you are happy with the final state for merging.

**Comment by jonatanschroeder** [comment]
> @jonatanschroeder I edited the news item so that it's now in a more "action oriented" order:
> 
> 1. What is this feature?
> 2. How do we configure it?
> 3. Why do we want it?
> 
> Please feel free to edit anything as always!

I added testing as an additional motivation. Otherwise looks good. And I don't mind if you change the news item, since you're listed as author anyway...

**Comment by mwest1066** [comment]
> I added testing as an additional motivation. Otherwise looks good. And I don't mind if you change the news item, since you're listed as author anyway...

Haha, so true 😄

Thanks!

**Pull request merged**
